### PR TITLE
Fix `TimeZoneInfo.Equals` and `TimeZoneInfo.HasSameRules`

### PR DESCRIPTION
**Pull request by RenderMichael**
Based on https://github.com/dotnet/runtime/pull/88641#issuecomment-1632932583

On Linux, different UTC aliases for produce TimeZoneInfos which have all the same rules, but some have `_adjustmentRules == null` and others `_adjustmentRules == []`. The corresponding API `GetAdjustmentRules()` returns an empty array if the underlying value is null:
https://github.com/dotnet/runtime/blob/a3c21b95c91012df3e2572a1331ed7cc1f2679e6/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.cs#L140-L146
https://github.com/dotnet/runtime/blob/a3c21b95c91012df3e2572a1331ed7cc1f2679e6/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Win32.cs#L81-L87
Which implies that these two values are equivalent.

### CODE REVIEW

File: src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs [hunk 1]
```diff
@@ -2446,6 +2446,16 @@ public static void UtcAliases_MapToUtc()
             }
         }
 
+        [Fact]
+        public static void UtcAliases_AreEqualToUtc()
```
Comment (tarekgh) [inline src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:2450]: >public static void UtcAliases_AreEqualToUtc() [](http://example.com/codeflow?start=8&length=45)

need to use `[PlatformSpecific(~TestPlatforms.Windows & ~TestPlatforms.Browser)]`?

File: src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs [hunk 2]
```diff
@@ -2446,6 +2446,16 @@ public static void UtcAliases_MapToUtc()
             }
         }
```
Comment (tarekgh) [inline src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:2447]: This test looks redundant now.

File: src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs [hunk 3]
```diff
@@ -2446,6 +2446,16 @@ public static void UtcAliases_MapToUtc()
...
 
+        [Fact]
+        public static void UtcAliases_AreEqualToUtc()
+        {
+            foreach (string alias in s_UtcAliases)
+            {
+                TimeZoneInfo actualUtc = TimeZoneInfo.FindSystemTimeZoneById(alias);
+                Assert.Equal(TimeZoneInfo.Utc, actualUtc);
```
Comment (tarekgh) [inline src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:2455]: >Assert.Equal(TimeZoneInfo.Utc, actualUtc); [](http://example.com/codeflow?start=16&length=42)

Could you please add 

```C#
Assert.Equal(actualUtc, TimeZoneInfo.Utc);
```

To ensure the order hold the right equality too.

File: src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs (hunk 1 above)
Comment (RenderMichael) [reply src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:2450]: As I linked in the description, `GetAdjustmentRules()` on windows also has logic to return empty array when the  underlying is null. This test is relevant (and should pass) on Windows

Comment (tarekgh) [reply src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:2450]: It is possible to run on Windows machines that do not have ICU library. At that time, you cannot create the time zone objects listed in `s_UtcAliases`.

Comment (RenderMichael) [reply src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:2450]: maybe we can do
``` csharp
foreach (string alias in s_UtcAliases)
{
    try
    {
        TimeZoneInfo actualUtc = TimeZoneInfo.FindSystemTimeZoneById(alias);
        // ...
    }
    catch (TimeZoneNotFoundException) { }
}
```

so if the timezones can be found, they are tested

Comment (RenderMichael) [reply src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:2450]: This is what the test data in `SystemTimeZonesTestData()` does
https://github.com/dotnet/runtime/blob/c88dbd3f943b8206b4821aaf10ddf3c216100942/src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs#L2322-L2330

Comment (tarekgh) [reply src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:2450]: It is better to use `[ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsIcuGlobalization))]` on the method.

**Comment by RenderMichael** [comment]
Build failure is https://github.com/dotnet/runtime/issues/88868, PTAL @tarekgh

**Review by tarekgh — CHANGES_REQUESTED** [review]

**Comment by tarekgh** [comment]
@RenderMichael all failures are known unrelated issues https://github.com/dotnet/runtime/pull/88869/checks?check_run_id=15059873081. Please try to look at the suggestion https://github.com/dotnet/runtime/pull/88869#discussion_r1264197943.

**Comment by RenderMichael** [comment]
@tarekgh Glad there's a way to express the test's prerequisites in a less ad-hoc way! Ready for review

**Review by tarekgh — APPROVED** [review]

**Pull request merged**